extensions/stackdriver/edges/edge_reporter.cc

```diff
@@ -0,0 +1,148 @@
+  current_request_ = std::make_unique<ReportTrafficAssertionsRequest>();
+
+  const auto& project_id =
+      local_node_info.platform_metadata().at(Common::kGCPProjectKey);
+  current_request_->set_parent("projects/" + project_id);
+
+  // TODO(dougreid): figure out how to get the real mesh uid here.
+  // Using: //.../projects/<project_id>/<location>/meshes/<cluster> as a
+  // placeholder.
+  std::string location =
+      local_node_info.platform_metadata().at(Common::kGCPClusterLocationKey);
+  std::string cluster =
+      local_node_info.platform_metadata().at(Common::kGCPClusterNameKey);
+  std::string mesh_uid = "unknown";
+  absl::StrAppend(current_request_->mutable_mesh_uid(),
+                  "//cloudresourcemanager.googleapis.com/projects/", project_id,
+                  "/", location, "/meshes/", cluster);
+
+  instanceFromMetadata(local_node_info, &node_instance_);
+
+  edges_client_ = std::move(edges_client);
+};
+
+// ONLY inbound
+void EdgeReporter::addEdge(const ::Wasm::Common::RequestInfo& request_info,
+                           const ::wasm::common::NodeInfo& peer_node_info) {
+  auto peer = current_peers_.find(peer_node_info.node_key());
+  if (peer != current_peers_.end()) {
+    return;
+  }
+  current_peers_.emplace(peer_node_info.node_key());
+
+  auto* traffic_assertions = current_request_->mutable_traffic_assertions();
+  auto* edge = traffic_assertions->Add();
+
+  edge->set_destination_service_name(request_info.destination_service_host);
+  edge->set_destination_service_namespace(node_instance_.workload_namespace());
+  instanceFromMetadata(peer_node_info, edge->mutable_source());
+  edge->mutable_destination()->CopyFrom(node_instance_);
+
+  auto protocol = request_info.request_protocol;
+  if (protocol == "http" || protocol == "HTTP") {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_HTTP);
+  } else if (protocol == "https" || protocol == "HTTPS") {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_HTTPS);
+  } else if (protocol == "grpc" || protocol == "GRPC") {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_GRPC);
+  } else {
+    edge->set_protocol(TrafficAssertion_Protocol_PROTOCOL_TCP);
+  }
+
+  if (current_request_->traffic_assertions_size() >
+      max_assertions_per_request_) {
+    flush();
+  }
+};  // namespace Edges
+
+void EdgeReporter::reportEdges() {
+  flush();
+  for (auto& req : request_queue_) {
+    edges_client_->reportTrafficAssertions(std::move(req));
+  }
+  request_queue_.clear();
+}
+
+void EdgeReporter::flush() {
+  if (current_request_->traffic_assertions_size() == 0) {
+    return;
+  }
+
+  std::unique_ptr<ReportTrafficAssertionsRequest> next =
+      std::make_unique<ReportTrafficAssertionsRequest>();
+  next->set_parent(current_request_->parent());
+  next->set_mesh_uid(current_request_->mesh_uid());
+
+  current_peers_.clear();
+  current_request_.swap(next);
+  *next->mutable_timestamp() = TimeUtil::GetCurrentTime();
```

Comment:
Just switching causes seg faults when used in tests (see below), so I've made a TimestampFn that can be passed in and set a default to use the wasm API.

```
[2019-10-08 17:09:55.505][14][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:81] Caught Segmentation fault, suspect faulting address 0x0
[2019-10-08 17:09:55.505][14][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:69] Backtrace (use tools/stack_decode.py to get line numbers):
[2019-10-08 17:09:55.514][14][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #0: Envoy::SignalAction::sigHandler() [0x5646547c2f32]
[2019-10-08 17:09:55.514][14][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #1: __restore_rt [0x7f6d4be8b3a0]
[2019-10-08 17:09:55.522][14][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #2: Envoy::Extensions::Common::Wasm::Null::Plugin::proxy_getCurrentTimeNanoseconds() [0x56465349f8c8]
[2019-10-08 17:09:55.530][14][critical][backtrace] [bazel-out/k8-fastbuild/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:73] #3: Envoy::Extensions::Common::Wasm::Null::Plugin::getCurrentTimeNanoseconds() [0x5646534a05a4]
```

Reply:
Ah, it is because the wasm runtime context is empty. Any calls from filter/modules into envoy host needs to attach to a runtime context: https://github.com/envoyproxy/envoy-wasm/blob/724328c7d84bae38d9e4aa7047121a949a6fd511/source/extensions/common/wasm/wasm.cc#L42

Reply:
There is no easy way to add unit test for method using WASM APIs, there is no mock provided. Let's switch back to `TimeUtil::GetCurrentTime()` for now and worry about wasm later.

Reply:
nvm, I understand your change better now. and yeah it is viable 👍